### PR TITLE
Added LogicalNotOperatorsWithSuccessorSpaceFixer

### PR DESCRIPTION
**Pull request by phansys**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets |
| License       | MIT
| Doc PR        |

This rule is intended to be used within Laravel.

### CODE REVIEW

File: Symfony/CS/Tests/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixerTest.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        return array(
+            array(
+                '<?php $i = 0; $i++; $foo = ! false || (! true || ! ! false);',
+                '<?php $i = 0; $i++; $foo = !false || (!true || !!false);',
+            ),
+            array(
+                '<?php $i = 0; $i--; $foo = ! false || ($i && ! true);',
+                '<?php $i = 0; $i--; $foo = !false || ($i && !true);',
+            ),
```
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixerTest.php]: Is `if (!    $foo) {}` correct ?

**Comment by keradus** [comment]
Does this fixer is in conflict with `UnaryOperatorsSpacesFixer` ? `UnaryOperatorsSpacesFixer` fixers also `- 5` => `-5`

File: Symfony/CS/Tests/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixerTest.php (hunk 1 above)
Comment (phansys) [inline Symfony/CS/Tests/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixerTest.php]: Yes @keradus, this fixer ensures at least a space is following the ```!``` operator.

**Comment by phansys** [comment]
I guess so @keradus, should I work with ```getPriority()```?

**Comment by phansys** [comment]
I added a [new commit](https://github.com/phansys/PHP-CS-Fixer/commit/007101d2c0805255d78462351641ff34f9040ce6) with that change. If it is valid, I'll squash later and backport the same update for #1303.

**Comment by keradus** [comment]
Then I understand it's not a conflicting fixer and `- 5` should be changed into `-5` for Laravel?
Then also please add tests [here](https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/1.10/Symfony/CS/Tests/FixerTest.php#L158)

**Comment by phansys** [comment]
Done @keradus.

File: Symfony/CS/Tests/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixerTest.php (hunk 1 above)
Comment (keradus) [inline Symfony/CS/Tests/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixerTest.php]: ok, sorry, github short code block trim spaces, it was like:
```php
if (!    $foo) {}
```

And if multiple spaces are allowed as you said - please add a case to prove they are not modified.

-------------------

Also, what about:
```php
if (!/*foo*/$foo) {}
```
and
```php
if (!/*foo*/      $foo) {}
```
?

File: Symfony/CS/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixer.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+namespace Symfony\CS\Fixer\Contrib;
+
+use Symfony\CS\AbstractFixer;
+use Symfony\CS\Tokenizer\Token;
+use Symfony\CS\Tokenizer\Tokens;
+
+/**
+ * @author Javier Spagnoletti <phansys@gmail.com>
+ */
+class LogicalNotOperatorsWithSuccessorSpaceFixer extends AbstractFixer
```
Comment (keradus) [inline Symfony/CS/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixer.php]: Please mark that class as final

**Comment by keradus** [comment]
> I added a new commit with that change. If it is valid, I'll squash later and backport the same update for #1303.

Yeah, it's great.
BTW, there is no need to squash it yourself.

File: Symfony/CS/Tests/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixerTest.php (hunk 1 above)
Comment (phansys) [inline Symfony/CS/Tests/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixerTest.php]: @keradus, tests added for these cases.

File: Symfony/CS/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixer.php (hunk 2 above)
Comment (phansys) [inline Symfony/CS/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixer.php]: Done.

File: Symfony/CS/Tests/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixerTest.php (hunk 1 above)
Comment (GrahamCampbell) [inline Symfony/CS/Tests/Fixer/Contrib/LogicalNotOperatorsWithSuccessorSpaceFixerTest.php]: > this fixer ensures at least a space is following the ! operator.

I'd prefer it if we fixed to force `1` space.

**Comment by GrahamCampbell** [comment]
Having fixer to ensure there's a least one space feels a bit pointless. Surely we should ensure there's exactly one space.

**Comment by GrahamCampbell** [comment]
That's the use case I'd want from the fixer, and so far, I'm the only one who's said I'd actually use it.

**Comment by GrahamCampbell** [comment]
That, is laravel/framework will use this fixer, and we want exactly one space.

**Comment by phansys** [comment]
@GrahamCampbell, updated.

**Comment by phansys** [comment]
Updated according your comment @keradus. Please tell me if that is the right way to achive what you suggested.

**Comment by keradus** [comment]
:+1:

**Comment by keradus** [comment]
Please @phansys , could you resolve conflict ?

**Comment by keradus** [comment]
INFO: I will release 1.10 after merge of this PR.

**Comment by GrahamCampbell** [comment]
Can we not just resolve the conflicts?

**Comment by keradus** [comment]
gh is playing with me during merging conflicted PR, that's why I have asked @phansys to do it.

**Comment by GrahamCampbell** [comment]
I would offer to do it, but I'm really busy right now, sorry. If this isn't done by Tuesday, I can do it. :)

**Comment by keradus** [comment]
Fixing it on separate PR isn't a problem, I just want to keep history of original PR. waiting for @phansys isn't a problem.

**Comment by keradus** [comment]
Thank you @phansys.

**Comment by phansys** [comment]
Sorry for delay @keradus, I wasn't at my usual location these days.
Thank you!

**Pull request merged**